server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearch.java

```diff
@@ -192,6 +204,81 @@ public void clear(String tenantId) {
         }
     }
 
+    @Override
+    public Trace getTraceFragment(String tenantId, String id) {
+        Span span = getSpan(tenantId, id);
+
+        if (span == null) {
+            return null;
+        }
+
+        InteractionNode interactionNode = spanToNode(span);
+        interactionNode.setNodes(recursiveTraceFragment(tenantId, span));
+
+        Trace trace = spanToTrace(span);
+        trace.getNodes().add(interactionNode);
+
+        return trace;
+    }
+
+
+    private List<Node> recursiveTraceFragment(String tenantId, Span parent) {
+        List<Span> spanChildren = getChildren(tenantId, SpanUniqueIdGenerator.toOriginal(parent));
+
+        if (spanChildren == null) {
+            return Collections.emptyList();
+        }
+
+        List<Node> nodes = new ArrayList<>();
+
+        for (Span child: spanChildren) {
+            InteractionNode node = spanToNode(child);
+            nodes.add(node);
+
+            if (node.getType() == NodeType.Component) {
```

Comment:
This needs to check for "!child.clientSpan()" - both Consumer and Component node types can have child nodes. It is only client spans that cannot have children.